plugins/utils/todo-comments.nix

```
@@ -1,290 +1,452 @@
-        inherit (cfg) keywords;
-        gui_style = cfg.guiStyle;
-        merge_keywords = cfg.mergeKeywords;
-        highlight = {
-          inherit (cfg.highlight)
-            multiline
-            before
-            keyword
-            after
-            exclude
-            ;
-          pattern = helpers.ifNonNull' cfg.highlight.pattern (helpers.mkRaw "[[${cfg.highlight.pattern}]]");
-          multiline_pattern = cfg.highlight.multilinePattern;
-          multiline_context = cfg.highlight.multilineContext;
-          comments_only = cfg.highlight.commentsOnly;
-          max_line_len = cfg.highlight.maxLineLen;
-        };
-        inherit (cfg) colors;
-        search = {
-          inherit (cfg.search) command args;
-          pattern = helpers.ifNonNull' cfg.search.pattern (
-            if isList cfg.search.pattern then
-              (map helpers.mkRaw cfg.search.pattern)
-            else
-              helpers.mkRaw "[[${cfg.search.pattern}]]"
-          );
-        };
-      } // cfg.extraOptions;
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-      extraPackages = [ cfg.ripgrepPackage ];
-      extraConfigLua = ''
-        require("todo-comments").setup${helpers.toLuaObject setupOptions}
-      '';
+  settingsExample = {
+    highlight = {
+      pattern = [
+        ".*<(KEYWORDS)\s*:"
+        ".*<(KEYWORDS)\s*"
+      ];
+    };
+  };
 
-      keymaps = flatten (
-        mapAttrsToList (
+  extraOptions = {
+    keymaps =
+      let
+        mkKeymapOption =
           optionName: funcName:
-          let
-            keymap = cfg.keymaps.${optionName};
-
-            cwd = optionalString (keymap.cwd != null) " cwd=${keymap.cwd}";
-            keywords = optionalString (keymap.keywords != null) " keywords=${keymap.keywords}";
-          in
-          optional (keymap != null) {
-            mode = "n";
-            inherit (keymap) key options;
-            action = ":${funcName}${cwd}${keywords}<CR>";
-          }
-        ) commands
-      );
-
-      # Automatically enable plugins if keymaps have been set
-      plugins = mkMerge [
-        (mkIf (cfg.keymaps.todoTrouble != null) { trouble.enable = true; })
-        (mkIf (cfg.keymaps.todoTelescope != null) { telescope.enable = true; })
-      ];
+          mkCompositeOption "Keymap settings for the `:${funcName}` function." {
+            key = mkOption {
+              type = types.str;
+              default = null;
+              description = "Key for the `${funcName}` function.";
+            };
+
+            cwd = mkOption {
+              type = types.nullOr types.str;
+              description = "Specify the directory to search for comments";
+              default = null;
+              example = "~/projects/foobar";
+            };
+
+            keywords = mkOption {
+              type =
+                with types;
+                transitionType str (string: if string != null then splitString "," string else null) (
```

I don't _think_ we need the null check in the coercion function?

If the value is null then it'll fail `types.str.check` and will end up being merged using the `nullOr` type anyway.

```suggestion
                transitionType str (splitString ",") (
```

---

It was required when i was doing transition with null old type, then I realized setting null as an old type caused it to trigger even when it wasn't needed. Now it's only attempting to coerce properly and it can be removed.